File: from-0601-to-0700/0623-add-one-row-to-tree/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func addOneRow(root *TreeNode, val int, depth int) *TreeNode {
    if depth == 1 {
        return &TreeNode{Val: val, Left: root}
    }
    var node *TreeNode
    q := []*TreeNode{root}
    lvl := 1
    for lvl < depth - 1 {
        size := len(q)
        for i := 0; i < size; i ++ {
            node := q[0]
            q = q[1:]
            if node.Left != nil {
                q = append(q, node.Left)
            }
            if node.Right != nil {
                q = append(q, node.Right)
            }
        }
        lvl += 1
    }

    for len(q) > 0 {
        node = q[0]
        q = q[1:]
        node.Left = &TreeNode{Val: val, Left: node.Left}
        node.Right = &TreeNode{Val: val, Right: node.Right}
    }
    return root
}

File: from-0601-to-0700/0623-add-one-row-to-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def addOneRow(self, root: Optional[TreeNode], val: int, depth: int) -> Optional[TreeNode]:
        if depth == 1:
            return TreeNode(val=val, left=root)
            
        def dfs(node: TreeNode, cur_depth: int) -> None:
            nonlocal depth, val
            if not node or cur_depth > depth:
                return
            
            if cur_depth == depth - 1:
                node.left = TreeNode(val=val, left=node.left)
                node.right = TreeNode(val=val, right=node.right)
                return
                
            dfs(node.left, cur_depth + 1)
            dfs(node.right, cur_depth + 1)
            
        dfs(root, 1)
        
        return root
    
#         if depth == 1:
#                 return TreeNode(val=val, left=root)

#         from queue import SimpleQueue
#         q = SimpleQueue()
#         level = 1
#         q.put(root)
#         while level < depth-1:
#             for _ in range(q.qsize()):
#                 node = q.get()
#                 if node.left:
#                     q.put(node.left)
#                 if node.right:
#                     q.put(node.right)
#             level += 1

#         while not q.empty():
#             node = q.get()
#             node.left = TreeNode(val=val, left=node.left)
#             node.right = TreeNode(val=val, right=node.right)

#         return root